<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>trackViewer Vignette: plot interaction data • trackViewer</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="trackViewer Vignette: plot interaction data">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">trackViewer</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.41.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/trackViewer.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/changeTracksStyles.html">trackViewer Vignette: change the track styles</a>
    </li>
    <li>
      <a href="../articles/dandelionPlot.html">trackViewer Vignette: dandelionPlot</a>
    </li>
    <li>
      <a href="../articles/lollipopPlot.html">trackViewer Vignette: lollipopPlot</a>
    </li>
    <li>
      <a href="../articles/plotInteractionData.html">trackViewer Vignette: plot interaction data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>trackViewer Vignette: plot interaction data</h1>
                        <h4 data-toc-skip class="author">Jianhong Ou,
Lihua Julie Zhu</h4>
            
            <h4 data-toc-skip class="date">5 September 2024</h4>
      

      <div class="hidden name"><code>plotInteractionData.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Visualize chromatin interactions along with annotation as track
      layers. The interactions can be compared by back to back heatmaps.
      The interactions can be plot as heatmap and links.</p>
    </div>
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The chromatin interactions is involved in precise quantitative and
spatiotemporal control of gene expression. The development of
high-throughput experimental techniques, such as HiC-seq, HiCAR-seq, and
InTAC-seq, for analyzing both the higher-order structure of chromatin
and the interactions between protein and their nearby and remote
regulatory elements has been developed to reveal how gene expression is
controlled in genome-wide.</p>
<p>The interaction data will be saved in the format of paired genome
coordinates with the interaction score. The popular format are
<code>.validPairs</code>, <code>.hic</code>, and <code>.cool</code>. The
<code>trackViewer</code> package can be used to handle those data to
plot the heatmap or the interaction links.</p>
</div>
<div class="section level2">
<h2 id="plot-chromatin-interactions-data-in-linear-layout">Plot chromatin interactions data in linear layout<a class="anchor" aria-label="anchor" href="#plot-chromatin-interactions-data-in-linear-layout"></a>
</h2>
<p>Plot chromatin interactions tracks as heatmap.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">trackViewer</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">InteractionSet</span><span class="op">)</span></span>
<span><span class="va">gi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"nij.chr6.51120000.53200000.gi.rds"</span>, package<span class="op">=</span><span class="st">"trackViewer"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GInteractions object with 6 interactions and 1 metadata column:</span></span>
<span><span class="co">##       seqnames1           ranges1     seqnames2           ranges2 |     score</span></span>
<span><span class="co">##           &lt;Rle&gt;         &lt;IRanges&gt;         &lt;Rle&gt;         &lt;IRanges&gt; | &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]      chr6 51120000-51160000 ---      chr6 51120000-51160000 |   45.1227</span></span>
<span><span class="co">##   [2]      chr6 51120000-51160000 ---      chr6 51160000-51200000 |   35.0006</span></span>
<span><span class="co">##   [3]      chr6 51120000-51160000 ---      chr6 51200000-51240000 |   44.7322</span></span>
<span><span class="co">##   [4]      chr6 51120000-51160000 ---      chr6 51240000-51280000 |   29.3507</span></span>
<span><span class="co">##   [5]      chr6 51120000-51160000 ---      chr6 51280000-51320000 |   38.8417</span></span>
<span><span class="co">##   [6]      chr6 51120000-51160000 ---      chr6 51320000-51360000 |   31.7063</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   regions: 53 ranges and 0 metadata columns</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## hicexplorer:hicConvertFormat tool can be used to convert other formats into GInteractions</span></span>
<span><span class="co">## eg: hicConvertFormat -m mESC_rep.hic --inputFormat hic --outputFormat cool -o mESC_rep.mcool</span></span>
<span><span class="co">##     hicConvertFormat -m mESC_rep.mcool::resolutions/10000 --inputFormat cool --outputFormat ginteractions -o mESC_rep.ginteractions --resolutions 10000</span></span>
<span><span class="co">## please note that metadata:score is used for plot.</span></span>
<span><span class="va">gi</span><span class="op">$</span><span class="va">border_color</span> <span class="op">&lt;-</span> <span class="cn">NA</span> <span class="co">## highlight some regions</span></span>
<span><span class="va">gi</span><span class="op">$</span><span class="va">border_color</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-envirAPI.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span>, <span class="fl">20</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-envirAPI.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">7</span>, <span class="fl">20</span>, replace<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">## The TADs will be drawn as lines at points start(first), center point, end(second).</span></span>
<span><span class="va">tads</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/GInteractions-class.html" class="external-link">GInteractions</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr6"</span>, </span>
<span>          <span class="fu">IRanges</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">51130001</span>, <span class="fl">51130001</span>, <span class="fl">51450001</span>, <span class="fl">52210001</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">20000</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr6"</span>, </span>
<span>          <span class="fu">IRanges</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">51530001</span>, <span class="fl">52170001</span>, <span class="fl">52210001</span>, <span class="fl">53210001</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">20000</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">range</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr6"</span>, <span class="fu">IRanges</span><span class="op">(</span><span class="fl">51120000</span>, <span class="fl">53200000</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">heatmap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gi2track.html">gi2track</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span></span>
<span><span class="va">ctcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"ctcf.sample.rds"</span>, package<span class="op">=</span><span class="st">"trackViewer"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/viewTracks.html">viewTracks</a></span><span class="op">(</span><span class="fu"><a href="../reference/trackList-class.html">trackList</a></span><span class="op">(</span><span class="va">ctcf</span>, <span class="va">heatmap</span>, heightDist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>, </span>
<span>           gr<span class="op">=</span><span class="va">range</span>, autoOptimizeStyle <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">## add TAD information</span></span>
<span><span class="fu"><a href="../reference/addInteractionAnnotation.html">addInteractionAnnotation</a></span><span class="op">(</span><span class="va">tads</span>, <span class="st">"heatmap"</span>, <span class="va">grid.lines</span>, gp<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>col<span class="op">=</span><span class="st">"#E69F00"</span>, lwd<span class="op">=</span><span class="fl">3</span>, lty<span class="op">=</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## add highlight interested regions</span></span>
<span><span class="va">gi_sub</span> <span class="op">&lt;-</span> <span class="va">gi</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">gi</span><span class="op">$</span><span class="va">score</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">gi_sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gi_sub</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/nearest-methods.html" class="external-link">distance</a></span><span class="op">(</span><span class="fu">first</span><span class="op">(</span><span class="va">gi_sub</span><span class="op">)</span>, <span class="fu">second</span><span class="op">(</span><span class="va">gi_sub</span><span class="op">)</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">200000</span><span class="op">]</span>, n<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">start</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html" class="external-link">regions</a></span><span class="op">(</span><span class="va">gi_sub</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">start</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html" class="external-link">regions</a></span><span class="op">(</span><span class="va">gi_sub</span><span class="op">)</span><span class="op">)</span><span class="op">-</span><span class="fl">40000</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">end</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html" class="external-link">regions</a></span><span class="op">(</span><span class="va">gi_sub</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">end</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html" class="external-link">regions</a></span><span class="op">(</span><span class="va">gi_sub</span><span class="op">)</span><span class="op">)</span><span class="op">+</span><span class="fl">40000</span></span>
<span><span class="fu"><a href="../reference/addInteractionAnnotation.html">addInteractionAnnotation</a></span><span class="op">(</span><span class="va">gi_sub</span>, <span class="st">"heatmap"</span>, <span class="va">grid.polygon</span>, gp<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>col<span class="op">=</span><span class="st">"red"</span>, lwd<span class="op">=</span><span class="fl">2</span>, lty<span class="op">=</span><span class="fl">2</span>, fill<span class="op">=</span><span class="cn">NA</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## add interesting anchor at giving coordinate.</span></span>
<span><span class="fu"><a href="../reference/addInteractionAnnotation.html">addInteractionAnnotation</a></span><span class="op">(</span><span class="fl">52900000</span>, <span class="st">"heatmap"</span>, gp<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>col<span class="op">=</span><span class="st">"blue"</span>, lwd<span class="op">=</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/addInteractionAnnotation.html">addInteractionAnnotation</a></span><span class="op">(</span><span class="op">-</span><span class="fl">52900000</span>, <span class="st">"heatmap"</span>, gp<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>col<span class="op">=</span><span class="st">"cyan"</span>, lwd<span class="op">=</span><span class="fl">3</span>, lty<span class="op">=</span><span class="fl">4</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="plotInteractionData_files/figure-html/plotback2back-1.png" width="700"></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## view the interaction data back to back.</span></span>
<span><span class="co">## Please make sure the data are normalized.</span></span>
<span><span class="va">gi2</span> <span class="op">&lt;-</span> <span class="va">gi</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">gi2</span><span class="op">$</span><span class="va">score</span> <span class="op">&lt;-</span> <span class="va">gi</span><span class="op">$</span><span class="va">score</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span>, sd <span class="op">=</span> <span class="fu">sd</span><span class="op">(</span><span class="va">gi</span><span class="op">$</span><span class="va">score</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">back2back</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gi2track.html">gi2track</a></span><span class="op">(</span><span class="va">gi</span>, <span class="va">gi2</span><span class="op">)</span></span>
<span><span class="co">## change the color</span></span>
<span><span class="fu"><a href="../reference/trackStyle-class.html">setTrackStyleParam</a></span><span class="op">(</span><span class="va">back2back</span>, <span class="st">"breaks"</span>, </span>
<span>                   <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from<span class="op">=</span><span class="fl">0</span>, to<span class="op">=</span><span class="fl">50</span>, by<span class="op">=</span><span class="fl">10</span><span class="op">)</span>, <span class="fl">200</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/trackStyle-class.html">setTrackStyleParam</a></span><span class="op">(</span><span class="va">back2back</span>, <span class="st">"color"</span>,</span>
<span>                   <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lightblue"</span>, <span class="st">"yellow"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## chang the lim of y-axis (by default, [0, 1])</span></span>
<span><span class="fu"><a href="../reference/trackStyle-class.html">setTrackStyleParam</a></span><span class="op">(</span><span class="va">back2back</span>, <span class="st">"ylim"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/viewTracks.html">viewTracks</a></span><span class="op">(</span><span class="fu"><a href="../reference/trackList-class.html">trackList</a></span><span class="op">(</span><span class="va">ctcf</span>, <span class="va">back2back</span>, heightDist<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span>,</span>
<span>           gr<span class="op">=</span><span class="va">range</span>, autoOptimizeStyle <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/addInteractionAnnotation.html">addInteractionAnnotation</a></span><span class="op">(</span><span class="va">tads</span>, <span class="st">"back2back"</span>, <span class="va">grid.lines</span>,</span>
<span>                         gp<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>col<span class="op">=</span><span class="st">"cyan"</span>, lwd<span class="op">=</span><span class="fl">3</span>, lty<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/addInteractionAnnotation.html">addInteractionAnnotation</a></span><span class="op">(</span><span class="op">-</span><span class="fl">52208000</span>, <span class="st">"back2back"</span>, gp<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>col<span class="op">=</span><span class="st">"blue"</span>, lwd<span class="op">=</span><span class="fl">3</span><span class="op">)</span>,</span>
<span>                         panel<span class="op">=</span><span class="st">"top"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/addInteractionAnnotation.html">addInteractionAnnotation</a></span><span class="op">(</span><span class="fl">51508000</span>, <span class="st">"back2back"</span>, gp<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>col<span class="op">=</span><span class="st">"gray"</span>, lwd<span class="op">=</span><span class="fl">3</span>, lty<span class="op">=</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>                         panel<span class="op">=</span><span class="st">"bottom"</span><span class="op">)</span></span></code></pre></div>
<p><img src="plotInteractionData_files/figure-html/plotback2back-2.png" width="700"></p>
<p>Plot chromatin interactions track as links.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/trackStyle-class.html">setTrackStyleParam</a></span><span class="op">(</span><span class="va">heatmap</span>, <span class="st">"tracktype"</span>, <span class="st">"link"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/trackStyle-class.html">setTrackStyleParam</a></span><span class="op">(</span><span class="va">heatmap</span>, <span class="st">"breaks"</span>, </span>
<span>                   <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from<span class="op">=</span><span class="fl">0</span>, to<span class="op">=</span><span class="fl">50</span>, by<span class="op">=</span><span class="fl">10</span><span class="op">)</span>, <span class="fl">200</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/trackStyle-class.html">setTrackStyleParam</a></span><span class="op">(</span><span class="va">heatmap</span>, <span class="st">"color"</span>,</span>
<span>                   <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lightblue"</span>, <span class="st">"yellow"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## filter the links to simulate the real data</span></span>
<span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/nearest-methods.html" class="external-link">distance</a></span><span class="op">(</span><span class="va">heatmap</span><span class="op">$</span><span class="va">dat</span>, <span class="va">heatmap</span><span class="op">$</span><span class="va">dat2</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">5e5</span> <span class="op">&amp;</span> <span class="va">heatmap</span><span class="op">$</span><span class="va">dat</span><span class="op">$</span><span class="va">score</span><span class="op">&gt;</span><span class="fl">20</span></span>
<span><span class="va">heatmap</span><span class="op">$</span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="va">heatmap</span><span class="op">$</span><span class="va">dat</span><span class="op">[</span><span class="va">keep</span><span class="op">]</span></span>
<span><span class="va">heatmap</span><span class="op">$</span><span class="va">dat2</span> <span class="op">&lt;-</span> <span class="va">heatmap</span><span class="op">$</span><span class="va">dat2</span><span class="op">[</span><span class="va">keep</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/viewTracks.html">viewTracks</a></span><span class="op">(</span><span class="fu"><a href="../reference/trackList-class.html">trackList</a></span><span class="op">(</span><span class="va">heatmap</span><span class="op">)</span>, gr<span class="op">=</span><span class="va">range</span>, autoOptimizeStyle <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="plotInteractionData_files/figure-html/plotLinks-1.png" width="576"></p>
<p>To import interactions data from “.hic” (reference to the script of
<a href="https://github.com/aidenlab/straw" class="external-link">hic-straw</a> and the <a href="https://www.cell.com/cms/10.1016/j.cels.2016.07.002/attachment/ce39448c-9a11-4b4e-a03f-45882b7b1d9d/mmc2.xlsx" class="external-link">documentation</a>).
The function <code>importGInteractions</code> (trackViewer
version&gt;=1.27.6) can be used to import data from <code>.hic</code>
format file.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test_chr22.hic"</span>, package <span class="op">=</span> <span class="st">"trackViewer"</span>,</span>
<span>                    mustWork<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="va">.Platform</span><span class="op">$</span><span class="va">OS.type</span><span class="op">!=</span><span class="st">"windows"</span><span class="op">)</span><span class="op">{</span></span>
<span><span class="fu"><a href="../reference/importGInteractions.html">importGInteractions</a></span><span class="op">(</span>file<span class="op">=</span><span class="va">hic</span>, format<span class="op">=</span><span class="st">"hic"</span>,</span>
<span>                    ranges<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"22"</span>, <span class="fu">IRanges</span><span class="op">(</span><span class="fl">50000000</span>, <span class="fl">100000000</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                    out <span class="op">=</span> <span class="st">"GInteractions"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<pre><code><span><span class="co">## GInteractions object with 70 interactions and 1 metadata column:</span></span>
<span><span class="co">##        seqnames1           ranges1     seqnames2           ranges2 |     score</span></span>
<span><span class="co">##            &lt;Rle&gt;         &lt;IRanges&gt;         &lt;Rle&gt;         &lt;IRanges&gt; | &lt;numeric&gt;</span></span>
<span><span class="co">##    [1]        22 50000001-50100000 ---        22 50000001-50100000 |        26</span></span>
<span><span class="co">##    [2]        22 50000001-50100000 ---        22 50100001-50200000 |         2</span></span>
<span><span class="co">##    [3]        22 50100001-50200000 ---        22 50100001-50200000 |        22</span></span>
<span><span class="co">##    [4]        22 50100001-50200000 ---        22 50200001-50300000 |         7</span></span>
<span><span class="co">##    [5]        22 50200001-50300000 ---        22 50200001-50300000 |        31</span></span>
<span><span class="co">##    ...       ...               ... ...       ...               ... .       ...</span></span>
<span><span class="co">##   [66]        22 50400001-50500000 ---        22 51200001-51300000 |         1</span></span>
<span><span class="co">##   [67]        22 50500001-50600000 ---        22 51200001-51300000 |         2</span></span>
<span><span class="co">##   [68]        22 50800001-50900000 ---        22 51200001-51300000 |         2</span></span>
<span><span class="co">##   [69]        22 51100001-51200000 ---        22 51200001-51300000 |         3</span></span>
<span><span class="co">##   [70]        22 51200001-51300000 ---        22 51200001-51300000 |         5</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   regions: 13 ranges and 0 metadata columns</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>Another widely used genomic interaction data format is
<code>.cool</code> files and the <a href="ftp://cooler.csail.mit.edu/coolers">cooler index</a> contains
analyzed HiC data for hg19 and mm9 from many different sources. Those
files can be used as data resources for visualizations and annotations
(see <a href="https://rdrr.io/bioc/ChIPpeakAnno/man/findEnhancers.html" class="external-link">ChIPpeakAnno::findEnhancers</a>).
The <code>importGInteractions</code> function can also be used to import
data from <code>.cool</code> format (trackViewer
version&gt;=1.27.6).</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cool</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test.mcool"</span>, package <span class="op">=</span> <span class="st">"trackViewer"</span>,</span>
<span>                     mustWork<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/importGInteractions.html">importGInteractions</a></span><span class="op">(</span>file<span class="op">=</span><span class="va">cool</span>, format<span class="op">=</span><span class="st">"cool"</span>,</span>
<span>                    resolution <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                    ranges<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="fu">IRanges</span><span class="op">(</span><span class="fl">10</span>, <span class="fl">28</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                    out <span class="op">=</span> <span class="st">"GInteractions"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>R version 4.4.1 (2024-06-14) Platform: x86_64-pc-linux-gnu Running
under: Ubuntu 22.04.4 LTS</p>
<p>Matrix products: default BLAS:
/usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 LAPACK:
/usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;
LAPACK version 3.10.0</p>
<p>locale: [1] LC_CTYPE=en_US.UTF-8 LC_NUMERIC=C<br>
[3] LC_TIME=en_US.UTF-8 LC_COLLATE=en_US.UTF-8<br>
[5] LC_MONETARY=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8<br>
[7] LC_PAPER=en_US.UTF-8 LC_NAME=C<br>
[9] LC_ADDRESS=C LC_TELEPHONE=C<br>
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C</p>
<p>time zone: Etc/UTC tzcode source: system (glibc)</p>
<p>attached base packages: [1] grid stats4 stats graphics grDevices
utils datasets [8] methods base</p>
<p>other attached packages: [1] InteractionSet_1.33.0<br>
[2] SummarizedExperiment_1.35.1<br>
[3] MatrixGenerics_1.17.0<br>
[4] matrixStats_1.4.0<br>
[5] org.Hs.eg.db_3.19.1<br>
[6] TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2 [7]
GenomicFeatures_1.57.0<br>
[8] AnnotationDbi_1.67.0<br>
[9] Biobase_2.65.1<br>
[10] trackViewer_1.41.6<br>
[11] GenomicRanges_1.57.1<br>
[12] GenomeInfoDb_1.41.1<br>
[13] IRanges_2.39.2<br>
[14] S4Vectors_0.43.2<br>
[15] BiocGenerics_0.51.1</p>
<p>loaded via a namespace (and not attached): [1] strawr_0.0.92
RColorBrewer_1.1-3 rstudioapi_0.16.0<br>
[4] jsonlite_1.8.8 magrittr_2.0.3 rmarkdown_2.28<br>
[7] fs_1.6.4 BiocIO_1.15.2 zlibbioc_1.51.1<br>
[10] ragg_1.3.2 vctrs_0.6.5 memoise_2.0.1<br>
[13] Rsamtools_2.21.1 RCurl_1.98-1.16 base64enc_0.1-3<br>
[16] htmltools_0.5.8.1 S4Arrays_1.5.7 progress_1.2.3<br>
[19] curl_5.2.2 Rhdf5lib_1.27.0 rhdf5_2.49.0<br>
[22] SparseArray_1.5.31 Formula_1.2-5 sass_0.4.9<br>
[25] bslib_0.8.0 htmlwidgets_1.6.4 desc_1.4.3<br>
[28] Gviz_1.49.0 httr2_1.0.3 cachem_1.1.0<br>
[31] GenomicAlignments_1.41.0 lifecycle_1.0.4 pkgconfig_2.0.3<br>
[34] Matrix_1.7-0 R6_2.5.1 fastmap_1.2.0<br>
[37] GenomeInfoDbData_1.2.12 digest_0.6.37 colorspace_2.1-1<br>
[40] textshaping_0.4.0 Hmisc_5.1-3 RSQLite_2.3.7<br>
[43] filelock_1.0.3 fansi_1.0.6 httr_1.4.7<br>
[46] abind_1.4-5 compiler_4.4.1 bit64_4.0.5<br>
[49] htmlTable_2.4.3 backports_1.5.0 BiocParallel_1.39.0<br>
[52] DBI_1.2.3 highr_0.11 biomaRt_2.61.3<br>
[55] rappdirs_0.3.3 DelayedArray_0.31.11 rjson_0.2.22<br>
[58] tools_4.4.1 foreign_0.8-87 nnet_7.3-19<br>
[61] glue_1.7.0 restfulr_0.0.15 rhdf5filters_1.17.0<br>
[64] checkmate_2.3.2 cluster_2.1.6 generics_0.1.3<br>
[67] gtable_0.3.5 BSgenome_1.73.0 ensembldb_2.29.1<br>
[70] data.table_1.16.0 hms_1.1.3 xml2_1.3.6<br>
[73] utf8_1.2.4 XVector_0.45.0 pillar_1.9.0<br>
[76] stringr_1.5.1 dplyr_1.1.4 BiocFileCache_2.13.0<br>
[79] lattice_0.22-6 deldir_2.0-4 rtracklayer_1.65.0<br>
[82] bit_4.0.5 biovizBase_1.53.0 tidyselect_1.2.1<br>
[85] Biostrings_2.73.1 knitr_1.48 gridExtra_2.3<br>
[88] ProtGenerics_1.37.1 xfun_0.47 stringi_1.8.4<br>
[91] UCSC.utils_1.1.0 lazyeval_0.2.2 yaml_2.3.10<br>
[94] evaluate_0.24.0 codetools_0.2-20 interp_1.1-6<br>
[97] tibble_3.2.1 BiocManager_1.30.25 cli_3.6.3<br>
[100] rpart_4.1.23 systemfonts_1.1.0 munsell_0.5.1<br>
[103] jquerylib_0.1.4 Rcpp_1.0.13 dichromat_2.0-0.1<br>
[106] grImport_0.9-7 dbplyr_2.5.0 png_0.1-8<br>
[109] XML_3.99-0.17 parallel_4.4.1 pkgdown_2.1.0<br>
[112] ggplot2_3.5.1 blob_1.2.4 prettyunits_1.2.0<br>
[115] latticeExtra_0.6-30 jpeg_0.1-10 AnnotationFilter_1.29.0 [118]
bitops_1.0-8 txdbmaker_1.1.1 VariantAnnotation_1.51.0 [121] scales_1.3.0
crayon_1.5.3 BiocStyle_2.33.1<br>
[124] rlang_1.1.4 KEGGREST_1.45.1</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jianhong Ou, Julie Lihua Zhu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
