<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>trackViewer Vignette: plot interaction data • trackViewer</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="trackViewer Vignette: plot interaction data">
<meta property="og:description" content="trackViewer">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">trackViewer</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.27.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/trackViewer.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/changeTracksStyles.html">trackViewer Vignette: change the track styles</a>
    </li>
    <li>
      <a href="../articles/dandelionPlot.html">trackViewer Vignette: dandelionPlot</a>
    </li>
    <li>
      <a href="../articles/lollipopPlot.html">trackViewer Vignette: lollipopPlot</a>
    </li>
    <li>
      <a href="../articles/plotInteractionData.html">trackViewer Vignette: plot interaction data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="plotInteractionData_files/header-attrs-2.7/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>trackViewer Vignette: plot interaction data</h1>
                        <h4 class="author">Jianhong Ou, Lihua Julie Zhu</h4>
            
            <h4 class="date">4 April 2021</h4>
      
      
      <div class="hidden name"><code>plotInteractionData.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Visualize chromatin interactions along with annotation as track layers. The interactions can be compared by back to back heatmaps. The interactions can be plot as heatmap and links.</p>
    </div>
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>The chromatin interactions is involved in precise quantitative and spatiotemporal control of gene expression. The development of high-throughput experimental techniques, such as HiC-seq, HiCAR-seq, and InTAC-seq, for analyzing both the higher-order structure of chromatin and the interactions between protein and their nearby and remote regulatory elements has been developed to reveal how gene expression is controlled in genome-wide.</p>
<p>The interaction data will be saved in the format of paired genome coordinates with the interaction score. The popular format are <code>.validPairs</code>, <code>.hic</code>, and <code>.cool</code>. The <code>trackViewer</code> package can be used to handle those data to plot the heatmap or the interaction links.</p>
</div>
<div id="plot-chromatin-interactions-data" class="section level1">
<h1 class="hasAnchor">
<a href="#plot-chromatin-interactions-data" class="anchor"></a>Plot chromatin interactions data</h1>
<p>Plot chromatin interactions tracks as heatmap.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">trackViewer</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">InteractionSet</span><span class="op">)</span>
<span class="va">gi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"nij.chr6.51120000.53200000.gi.rds"</span>, package<span class="op">=</span><span class="st">"trackViewer"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span></code></pre></div>
<pre><code>## GInteractions object with 6 interactions and 1 metadata column:
##       seqnames1           ranges1     seqnames2           ranges2 |     score
##           &lt;Rle&gt;         &lt;IRanges&gt;         &lt;Rle&gt;         &lt;IRanges&gt; | &lt;numeric&gt;
##   [1]      chr6 51120000-51160000 ---      chr6 51120000-51160000 |   45.1227
##   [2]      chr6 51120000-51160000 ---      chr6 51160000-51200000 |   35.0006
##   [3]      chr6 51120000-51160000 ---      chr6 51200000-51240000 |   44.7322
##   [4]      chr6 51120000-51160000 ---      chr6 51240000-51280000 |   29.3507
##   [5]      chr6 51120000-51160000 ---      chr6 51280000-51320000 |   38.8417
##   [6]      chr6 51120000-51160000 ---      chr6 51320000-51360000 |   31.7063
##   -------
##   regions: 53 ranges and 0 metadata columns
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## hicexplorer:hicConvertFormat tool can be used to convert other formats into GInteractions</span>
<span class="co">## eg: hicConvertFormat -m mESC_rep.hic --inputFormat hic --outputFormat cool -o mESC_rep.mcool</span>
<span class="co">##     hicConvertFormat -m mESC_rep.mcool::resolutions/10000 --inputFormat cool --outputFormat ginteractions -o mESC_rep.ginteractions --resolutions 10000</span>
<span class="co">## please note that metadata:score is used for plot.</span>
<span class="va">range</span> <span class="op">&lt;-</span> <span class="fu">GRanges</span><span class="op">(</span><span class="st">"chr6"</span>, <span class="fu">IRanges</span><span class="op">(</span><span class="fl">51120000</span>, <span class="fl">53200000</span><span class="op">)</span><span class="op">)</span>
<span class="va">tr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gi2track.html">gi2track</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span>
<span class="va">ctcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"ctcf.sample.rds"</span>, package<span class="op">=</span><span class="st">"trackViewer"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#viewTracks(trackList(ctcf, tr, heightDist = c(1, 3)), </span>
<span class="co">#           gr=range, autoOptimizeStyle = TRUE)</span>
<span class="co">## view the interaction data back to back.</span>
<span class="co">## Please make sure the data are normalized.</span>
<span class="va">gi2</span> <span class="op">&lt;-</span> <span class="va">gi</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span>
<span class="va">gi2</span><span class="op">$</span><span class="va">score</span> <span class="op">&lt;-</span> <span class="va">gi</span><span class="op">$</span><span class="va">score</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span>, sd <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span><span class="op">(</span><span class="va">gi</span><span class="op">$</span><span class="va">score</span><span class="op">)</span><span class="op">)</span>
<span class="va">back2back</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gi2track.html">gi2track</a></span><span class="op">(</span><span class="va">gi</span>, <span class="va">gi2</span><span class="op">)</span>
<span class="co">## change the color</span>
<span class="fu"><a href="../reference/trackStyle-class.html">setTrackStyleParam</a></span><span class="op">(</span><span class="va">back2back</span>, <span class="st">"breaks"</span>, 
                   <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span>from<span class="op">=</span><span class="fl">0</span>, to<span class="op">=</span><span class="fl">50</span>, by<span class="op">=</span><span class="fl">10</span><span class="op">)</span>, <span class="fl">200</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/trackStyle-class.html">setTrackStyleParam</a></span><span class="op">(</span><span class="va">back2back</span>, <span class="st">"color"</span>,
                   <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"lightblue"</span>, <span class="st">"yellow"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/viewTracks.html">viewTracks</a></span><span class="op">(</span><span class="fu"><a href="../reference/trackList-class.html">trackList</a></span><span class="op">(</span><span class="va">ctcf</span>, <span class="va">back2back</span>, heightDist<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span>,
           gr<span class="op">=</span><span class="va">range</span>, autoOptimizeStyle <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="plotInteractionData_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<p>Plot chromatin interactions track as links.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/trackStyle-class.html">setTrackStyleParam</a></span><span class="op">(</span><span class="va">tr</span>, <span class="st">"tracktype"</span>, <span class="st">"link"</span><span class="op">)</span>
<span class="fu"><a href="../reference/trackStyle-class.html">setTrackStyleParam</a></span><span class="op">(</span><span class="va">tr</span>, <span class="st">"breaks"</span>, 
                   <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span>from<span class="op">=</span><span class="fl">0</span>, to<span class="op">=</span><span class="fl">50</span>, by<span class="op">=</span><span class="fl">10</span><span class="op">)</span>, <span class="fl">200</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/trackStyle-class.html">setTrackStyleParam</a></span><span class="op">(</span><span class="va">tr</span>, <span class="st">"color"</span>,
                   <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"lightblue"</span>, <span class="st">"yellow"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span>
<span class="co">## filter the links to simulate the real data</span>
<span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu">distance</span><span class="op">(</span><span class="va">tr</span><span class="op">$</span><span class="va">dat</span>, <span class="va">tr</span><span class="op">$</span><span class="va">dat2</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">5e5</span> <span class="op">&amp;</span> <span class="va">tr</span><span class="op">$</span><span class="va">dat</span><span class="op">$</span><span class="va">score</span><span class="op">&gt;</span><span class="fl">20</span>
<span class="va">tr</span><span class="op">$</span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="va">tr</span><span class="op">$</span><span class="va">dat</span><span class="op">[</span><span class="va">keep</span><span class="op">]</span>
<span class="va">tr</span><span class="op">$</span><span class="va">dat2</span> <span class="op">&lt;-</span> <span class="va">tr</span><span class="op">$</span><span class="va">dat2</span><span class="op">[</span><span class="va">keep</span><span class="op">]</span>
<span class="fu"><a href="../reference/viewTracks.html">viewTracks</a></span><span class="op">(</span><span class="fu"><a href="../reference/trackList-class.html">trackList</a></span><span class="op">(</span><span class="va">tr</span><span class="op">)</span>, gr<span class="op">=</span><span class="va">range</span>, autoOptimizeStyle <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="plotInteractionData_files/figure-html/unnamed-chunk-3-1.png" width="576"></p>
<p>To import interactions data from “.hic” (reference to the script of <a href="https://github.com/aidenlab/straw">hic-straw</a> and the <a href="https://www.cell.com/cms/10.1016/j.cels.2016.07.002/attachment/ce39448c-9a11-4b4e-a03f-45882b7b1d9d/mmc2.xlsx">documentation</a>). The function <code>importGInteractions</code> (trackViewer version&gt;=1.27.6) can be used to import data from <code>.hic</code> format file.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test_chr22.hic"</span>, package <span class="op">=</span> <span class="st">"trackViewer"</span>,
                    mustWork<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="../reference/importGInteractions.html">importGInteractions</a></span><span class="op">(</span>file<span class="op">=</span><span class="va">hic</span>, format<span class="op">=</span><span class="st">"hic"</span>,
                    ranges<span class="op">=</span><span class="fu">GRanges</span><span class="op">(</span><span class="st">"22"</span>, <span class="fu">IRanges</span><span class="op">(</span><span class="fl">50000000</span>, <span class="fl">100000000</span><span class="op">)</span><span class="op">)</span>,
                    out <span class="op">=</span> <span class="st">"GInteractions"</span><span class="op">)</span></code></pre></div>
<pre><code>## GInteractions object with 70 interactions and 1 metadata column:
##        seqnames1           ranges1     seqnames2   ranges2 |     score
##            &lt;Rle&gt;         &lt;IRanges&gt;         &lt;Rle&gt; &lt;IRanges&gt; | &lt;numeric&gt;
##    [1]        22 50000000-50100000 ---        22  50000000 |        26
##    [2]        22 50000000-50100000 ---        22  50100000 |         2
##    [3]        22 50100000-50200000 ---        22  50100000 |        22
##    [4]        22 50100000-50200000 ---        22  50200000 |         7
##    [5]        22 50200000-50300000 ---        22  50200000 |        31
##    ...       ...               ... ...       ...       ... .       ...
##   [66]        22 50400000-50500000 ---        22  51200000 |         1
##   [67]        22 50500000-50600000 ---        22  51200000 |         2
##   [68]        22 50800000-50900000 ---        22  51200000 |         2
##   [69]        22 51100000-51200000 ---        22  51200000 |         3
##   [70]        22 51200000-51300000 ---        22  51200000 |         5
##   -------
##   regions: 26 ranges and 0 metadata columns
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>Another widely used genomic interaction data format is <code>.cool</code> files and the <a href="ftp://cooler.csail.mit.edu/coolers">cooler index</a> contains analyzed HiC data for hg19 and mm9 from many different sources. Those files can be used as data resources for visualizations and annotations (see <a href="https://rdrr.io/bioc/ChIPpeakAnno/man/findEnhancers.html">ChIPpeakAnno::findEnhancers</a>). The <code>importGInteractions</code> function can also be used to import data from <code>.cool</code> format (trackViewer version&gt;=1.27.6).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cool</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test.mcool"</span>, package <span class="op">=</span> <span class="st">"trackViewer"</span>,
                     mustWork<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="../reference/importGInteractions.html">importGInteractions</a></span><span class="op">(</span>file<span class="op">=</span><span class="va">cool</span>, format<span class="op">=</span><span class="st">"cool"</span>,
                    resolution <span class="op">=</span> <span class="fl">2</span>,
                    ranges<span class="op">=</span><span class="fu">GRanges</span><span class="op">(</span><span class="st">"chr1"</span>, <span class="fu">IRanges</span><span class="op">(</span><span class="fl">10</span>, <span class="fl">28</span><span class="op">)</span><span class="op">)</span>,
                    out <span class="op">=</span> <span class="st">"GInteractions"</span><span class="op">)</span></code></pre></div>
<p>Different from most of the available tools, plotGInteractions try to plot the data with the 2D structure. The nodes indicate the region with interactions and the edges indicates the interactions. The size of the nodes are relative to the width of the region. The features could be the enhancers, promoters or genes. The enhancer and promoter are shown as points with symbol 11 and 13.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">InteractionSet</span><span class="op">)</span>
<span class="va">gi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"gi.rds"</span>, package<span class="op">=</span><span class="st">"trackViewer"</span><span class="op">)</span><span class="op">)</span>
<span class="va">range</span> <span class="op">&lt;-</span> <span class="fu">GRanges</span><span class="op">(</span><span class="st">"chr2"</span>, <span class="fu">IRanges</span><span class="op">(</span><span class="fl">234500000</span>, <span class="fl">235000000</span><span class="op">)</span><span class="op">)</span>
<span class="va">feature.gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressMessages</a></span><span class="op">(</span><span class="fu">genes</span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span><span class="op">)</span>
<span class="va">feature.gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/overlaps.html">subsetByOverlaps</a></span><span class="op">(</span><span class="va">feature.gr</span>, <span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">regions</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span><span class="op">)</span>
<span class="va">feature.gr</span><span class="op">$</span><span class="va">col</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">7</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">feature.gr</span><span class="op">)</span>, replace<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="va">feature.gr</span><span class="op">$</span><span class="va">type</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"promoter"</span>, <span class="st">"enhancer"</span>, <span class="st">"gene"</span><span class="op">)</span>, 
                          <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">feature.gr</span><span class="op">)</span>, replace<span class="op">=</span><span class="cn">TRUE</span>, 
                          prob<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.2</span>, <span class="fl">0.7</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotGInteractions.html">plotGInteractions</a></span><span class="op">(</span><span class="va">gi</span>, <span class="va">range</span>, <span class="va">feature.gr</span><span class="op">)</span></code></pre></div>
<p><img src="plotInteractionData_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h1>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>R Under development (unstable) (2021-03-18 r80099) Platform: x86_64-pc-linux-gnu (64-bit) Running under: Ubuntu 20.04.2 LTS</p>
<p>Matrix products: default BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so</p>
<p>locale: [1] LC_CTYPE=en_US.UTF-8 LC_NUMERIC=C<br>
[3] LC_TIME=en_US.UTF-8 LC_COLLATE=en_US.UTF-8<br>
[5] LC_MONETARY=en_US.UTF-8 LC_MESSAGES=C<br>
[7] LC_PAPER=en_US.UTF-8 LC_NAME=C<br>
[9] LC_ADDRESS=C LC_TELEPHONE=C<br>
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C</p>
<p>attached base packages: [1] grid parallel stats4 stats graphics grDevices utils<br>
[8] datasets methods base</p>
<p>other attached packages: [1] InteractionSet_1.19.0<br>
[2] SummarizedExperiment_1.21.1<br>
[3] MatrixGenerics_1.3.1<br>
[4] matrixStats_0.58.0<br>
[5] TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2 [6] GenomicFeatures_1.43.8<br>
[7] AnnotationDbi_1.53.1<br>
[8] Biobase_2.51.0<br>
[9] trackViewer_1.27.11<br>
[10] Rcpp_1.0.6<br>
[11] GenomicRanges_1.43.3<br>
[12] GenomeInfoDb_1.27.8<br>
[13] IRanges_2.25.6<br>
[14] S4Vectors_0.29.11<br>
[15] BiocGenerics_0.37.1</p>
<p>loaded via a namespace (and not attached): [1] colorspace_2.0-0 rjson_0.2.20 ellipsis_0.3.1<br>
[4] rprojroot_2.0.2 biovizBase_1.39.0 htmlTable_2.1.0<br>
[7] XVector_0.31.1 base64enc_0.1-3 fs_1.5.0<br>
[10] dichromat_2.0-0 rstudioapi_0.13 bit64_4.0.5<br>
[13] fansi_0.4.2 splines_4.1.0 cachem_1.0.4<br>
[16] knitr_1.31 Formula_1.2-4 Rsamtools_2.7.1<br>
[19] cluster_2.1.1 dbplyr_2.1.0 png_0.1-7<br>
[22] grImport_0.9-3 graph_1.69.0 BiocManager_1.30.12<br>
[25] compiler_4.1.0 httr_1.4.2 backports_1.2.1<br>
[28] lazyeval_0.2.2 assertthat_0.2.1 Matrix_1.3-2<br>
[31] fastmap_1.1.0 htmltools_0.5.1.1 prettyunits_1.1.1<br>
[34] tools_4.1.0 gtable_0.3.0 glue_1.4.2<br>
[37] GenomeInfoDbData_1.2.4 dplyr_1.0.5 rappdirs_0.3.3<br>
[40] pkgdown_1.6.1 rhdf5filters_1.3.4 vctrs_0.3.7<br>
[43] Biostrings_2.59.2 debugme_1.1.0 rtracklayer_1.51.5<br>
[46] xfun_0.22 stringr_1.4.0 lifecycle_1.0.0<br>
[49] ensembldb_2.15.2 restfulr_0.0.13 XML_3.99-0.6<br>
[52] zlibbioc_1.37.0 scales_1.1.1 BiocStyle_2.19.2<br>
[55] BSgenome_1.59.2 VariantAnnotation_1.37.1 ProtGenerics_1.23.7<br>
[58] ragg_1.1.2 hms_1.0.0 rhdf5_2.35.2<br>
[61] AnnotationFilter_1.15.0 RColorBrewer_1.1-2 yaml_2.2.1<br>
[64] curl_4.3 memoise_2.0.0 gridExtra_2.3<br>
[67] ggplot2_3.3.3 biomaRt_2.47.7 rpart_4.1-15<br>
[70] latticeExtra_0.6-29 stringi_1.5.3 RSQLite_2.2.5<br>
[73] highr_0.8 BiocIO_1.1.2 desc_1.3.0<br>
[76] plotrix_3.8-1 checkmate_2.0.0 filelock_1.0.2<br>
[79] BiocParallel_1.25.5 rlang_0.4.10 pkgconfig_2.0.3<br>
[82] systemfonts_1.0.1 bitops_1.0-6 evaluate_0.14<br>
[85] lattice_0.20-41 Rhdf5lib_1.13.4 purrr_0.3.4<br>
[88] GenomicAlignments_1.27.2 htmlwidgets_1.5.3 bit_4.0.4<br>
[91] tidyselect_1.1.0 magrittr_2.0.1 R6_2.5.0<br>
[94] generics_0.1.0 Hmisc_4.5-0 DelayedArray_0.17.10<br>
[97] DBI_1.1.1 pillar_1.5.1 foreign_0.8-81<br>
[100] survival_3.2-10 KEGGREST_1.31.1 RCurl_1.98-1.3<br>
[103] nnet_7.3-15 tibble_3.1.0 crayon_1.4.1<br>
[106] utf8_1.2.1 BiocFileCache_1.15.1 rmarkdown_2.7<br>
[109] jpeg_0.1-8.1 progress_1.2.2 data.table_1.14.0<br>
[112] Rgraphviz_2.35.0 blob_1.2.1 digest_0.6.27<br>
[115] textshaping_0.3.3 munsell_0.5.0 Gviz_1.35.5</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Jianhong Ou, Julie Lihua Zhu.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
