<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>trackViewer Vignette: overview • trackViewer</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="trackViewer Vignette: overview">
<meta property="og:description" content="trackViewer">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">trackViewer</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.27.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/trackViewer.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/changeTracksStyles.html">trackViewer Vignette: change the track styles</a>
    </li>
    <li>
      <a href="../articles/dandelionPlot.html">trackViewer Vignette: dandelionPlot</a>
    </li>
    <li>
      <a href="../articles/lollipopPlot.html">trackViewer Vignette: lollipopPlot</a>
    </li>
    <li>
      <a href="../articles/plotInteractionData.html">trackViewer Vignette: plot interaction data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="trackViewer_files/header-attrs-2.7/header-attrs.js"></script><script src="trackViewer_files/htmlwidgets-1.5.3/htmlwidgets.js"></script><script src="trackViewer_files/d3-4.4.0/d3.v4.min.js"></script><script src="trackViewer_files/saveSvgAsPng-0.0.1/saveSvgAsPng.js"></script><script src="trackViewer_files/browseTracks-binding-1.27.11/browseTracks.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>trackViewer Vignette: overview</h1>
                        <h4 class="author">Jianhong Ou, Lihua Julie Zhu</h4>
            
            <h4 class="date">4 April 2021</h4>
      
      
      <div class="hidden name"><code>trackViewer.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Visualize mapped reads along with annotation as track layers for NGS dataset such as ChIP-seq, RNA-seq, miRNA-seq, DNA-seq, SNPs and methylation data.</p>
    </div>
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>There are two packages available in Bioconductor for visualizing genomic data: <strong>rtracklayer</strong> and <strong>Gviz</strong>. <strong>rtracklayer</strong> provides an interface to genome browsers and associated annotation tracks. <strong>Gviz</strong> can be used to plot coverage and annotation tracks. <strong>TrackViewer</strong> is a lightweight visualization tool for generating interactive figures for publication. Not only can <strong>trackViewer</strong> be used to visualize coverage and annotation tracks, but it can also be employed to generate lollipop/dandelion plots that depict dense methylation/mutation/variant data to facilitate an integrative analysis of these multi-omics data. It leverages <strong>Gviz</strong> and <strong>rtracklayer</strong>, is easy to use, and has a low memory and cpu consumption. In addition, we implemented a web application of <strong>trackViewer</strong> leveraging <strong>Shiny</strong> package. The web application of <strong>trackViewer</strong> is available at <a href="https://github.com/jianhong/trackViewer.documentation/tree/master/trackViewerShinyApp" class="uri">https://github.com/jianhong/trackViewer.documentation/tree/master/trackViewerShinyApp</a>.</p>
<ul>
<li><p><a href="changeTracksStyles.html">Change the styles of the track plot</a></p></li>
<li><p><a href="lollipopPlot.html">Plot methylation/mutation/variant data by lollipop plot</a></p></li>
<li><p><a href="dandelionPlot.html">Plot high dense methylation/mutation/variant data by dandelion plot</a></p></li>
<li><p><a href="dandelionPlot.html">Plot genomic interaction data (HiC, cooler, etc) by viewTracks</a></p></li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ivanek/Gviz">Gviz</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">trackViewer</span><span class="op">)</span>
<span class="va">extdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"trackViewer"</span>,
                       mustWork<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu">GRanges</span><span class="op">(</span><span class="st">"chr11"</span>, <span class="fu">IRanges</span><span class="op">(</span><span class="fl">122929275</span>, <span class="fl">122930122</span><span class="op">)</span>, strand<span class="op">=</span><span class="st">"-"</span><span class="op">)</span>
<span class="va">fox2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importScore.html">importScore</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">extdata</span>, <span class="st">"fox2.bed"</span><span class="op">)</span>, format<span class="op">=</span><span class="st">"BED"</span>,
                    ranges<span class="op">=</span><span class="va">gr</span><span class="op">)</span>
<span class="va">fox2</span><span class="op">$</span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coverageGR.html">coverageGR</a></span><span class="op">(</span><span class="va">fox2</span><span class="op">$</span><span class="va">dat</span><span class="op">)</span>

<span class="fu"><a href="../reference/viewTracks.html">viewTracks</a></span><span class="op">(</span><span class="fu"><a href="../reference/trackList-class.html">trackList</a></span><span class="op">(</span><span class="va">fox2</span><span class="op">)</span>, gr<span class="op">=</span><span class="va">gr</span>, autoOptimizeStyle<span class="op">=</span><span class="cn">TRUE</span>, newpage<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>

<span class="va">dt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/DataTrack-class.html">DataTrack</a></span><span class="op">(</span>range<span class="op">=</span><span class="va">fox2</span><span class="op">$</span><span class="va">dat</span><span class="op">[</span><span class="fu">strand</span><span class="op">(</span><span class="va">fox2</span><span class="op">$</span><span class="va">dat</span><span class="op">)</span><span class="op">==</span><span class="st">"-"</span><span class="op">]</span> , 
                genome<span class="op">=</span><span class="st">"hg19"</span>, type<span class="op">=</span><span class="st">"hist"</span>, name<span class="op">=</span><span class="st">"fox2"</span>, 
                window<span class="op">=</span><span class="op">-</span><span class="fl">1</span>, chromosome<span class="op">=</span><span class="st">"chr11"</span>, 
                fill.histogram<span class="op">=</span><span class="st">"black"</span>, col.histogram<span class="op">=</span><span class="st">"NA"</span>,
                background.title<span class="op">=</span><span class="st">"white"</span>,
                col.frame<span class="op">=</span><span class="st">"white"</span>, col.axis<span class="op">=</span><span class="st">"black"</span>,
                col<span class="op">=</span><span class="st">"black"</span>, col.title<span class="op">=</span><span class="st">"black"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/plotTracks.html">plotTracks</a></span><span class="op">(</span><span class="va">dt</span>, from<span class="op">=</span><span class="fl">122929275</span>, to<span class="op">=</span><span class="fl">122930122</span>, strand<span class="op">=</span><span class="st">"-"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="trackViewer_files/figure-html/Gviz-1.png" alt="Plot data with **Gviz** and **trackViewer**. Please note that **trackViewer** can generate similar figure as **Gviz** with several lines of simple codes." width="768"><p class="caption">
Plot data with <strong>Gviz</strong> and <strong>trackViewer</strong>. Please note that <strong>trackViewer</strong> can generate similar figure as <strong>Gviz</strong> with several lines of simple codes.
</p>
</div>
<p><strong>trackViewer</strong> not only has the functionalities to produce the figures generated by <strong>Gviz</strong>, as shown in the Figure above, but also provides additional plotting styles as shown in the Figure below. The mimimalist design requires minimum input from the users while retaining the flexibility to change the output style easily.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu">GRanges</span><span class="op">(</span><span class="st">"chr1"</span>, <span class="fu">IRanges</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">6</span>, <span class="fl">10</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">6</span>, <span class="fl">12</span><span class="op">)</span><span class="op">)</span>, score<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span>
<span class="va">dt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/DataTrack-class.html">DataTrack</a></span><span class="op">(</span>range<span class="op">=</span><span class="va">gr</span>, data<span class="op">=</span><span class="st">"score"</span>, type<span class="op">=</span><span class="st">"hist"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/plotTracks.html">plotTracks</a></span><span class="op">(</span><span class="va">dt</span>, from<span class="op">=</span><span class="fl">2</span>, to<span class="op">=</span><span class="fl">11</span><span class="op">)</span>
<span class="va">tr</span> <span class="op">&lt;-</span> <span class="fu">new</span><span class="op">(</span><span class="st">"track"</span>, dat<span class="op">=</span><span class="va">gr</span>, type<span class="op">=</span><span class="st">"data"</span>, format<span class="op">=</span><span class="st">"BED"</span><span class="op">)</span>
<span class="fu"><a href="../reference/viewTracks.html">viewTracks</a></span><span class="op">(</span><span class="fu"><a href="../reference/trackList-class.html">trackList</a></span><span class="op">(</span><span class="va">tr</span><span class="op">)</span>, chromosome<span class="op">=</span><span class="st">"chr1"</span>, start<span class="op">=</span><span class="fl">2</span>, end<span class="op">=</span><span class="fl">11</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="trackViewer_files/figure-html/GvizLost-1.png" alt="Plot data with **Gviz** and **trackViewer**. Note that **trackViewer** is not only including more details but also showing all the data involved in the given range." width="768"><p class="caption">
Plot data with <strong>Gviz</strong> and <strong>trackViewer</strong>. Note that <strong>trackViewer</strong> is not only including more details but also showing all the data involved in the given range.
</p>
</div>
<p><strong>Gviz</strong> requires huge memory space to handle big wig files. To solve this problem, we rewrote the import function in <strong>trackViewer</strong> by importing the entire file first and parsing it later when plot. As a result, <strong>trackViewer</strong> decreases the import time from 180 min to 21 min and the memory cost from 10G to 5.32G for a half giga wig file (GSM917672).</p>
</div>
<div id="browse" class="section level1">
<h1 class="hasAnchor">
<a href="#browse" class="anchor"></a>Browse</h1>
<div id="steps-of-using-trackviewer" class="section level2">
<h2 class="hasAnchor">
<a href="#steps-of-using-trackviewer" class="anchor"></a>Steps of using trackViewer </h2>
<div id="step-1--import-data" class="section level3">
<h3 class="hasAnchor">
<a href="#step-1--import-data" class="anchor"></a>Step 1. Import data</h3>
<p>The function <strong>importScore</strong> is used to import BED, WIG, bedGraph or BigWig files. The function <strong>importBam</strong> is employed to import the bam files. Here is an example.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">trackViewer</span><span class="op">)</span>
<span class="va">extdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"trackViewer"</span>,
                       mustWork<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="va">repA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importScore.html">importScore</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">extdata</span>, <span class="st">"cpsf160.repA_-.wig"</span><span class="op">)</span>,
                    <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">extdata</span>, <span class="st">"cpsf160.repA_+.wig"</span><span class="op">)</span>,
                    format<span class="op">=</span><span class="st">"WIG"</span><span class="op">)</span>
<span class="co">## Because the wig file does not contain any strand info, </span>
<span class="co">## we need to set it manually.</span>
<span class="fu">strand</span><span class="op">(</span><span class="va">repA</span><span class="op">$</span><span class="va">dat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"-"</span>
<span class="fu">strand</span><span class="op">(</span><span class="va">repA</span><span class="op">$</span><span class="va">dat2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"+"</span></code></pre></div>
<p>The function <strong>coverageGR</strong> could be used to calculate the coverage after the data is imported.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fox2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importScore.html">importScore</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">extdata</span>, <span class="st">"fox2.bed"</span><span class="op">)</span>, format<span class="op">=</span><span class="st">"BED"</span>,
                    ranges<span class="op">=</span><span class="fu">GRanges</span><span class="op">(</span><span class="st">"chr11"</span>, <span class="fu">IRanges</span><span class="op">(</span><span class="fl">122830799</span>, <span class="fl">123116707</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coverageGR.html">coverageGR</a></span><span class="op">(</span><span class="va">fox2</span><span class="op">$</span><span class="va">dat</span><span class="op">)</span>
<span class="co">## We can split the data by strand into two different track channels</span>
<span class="co">## Here, we set the dat2 slot to save the negative strand info. </span>
 
<span class="va">fox2</span><span class="op">$</span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="va">dat</span><span class="op">[</span><span class="fu">strand</span><span class="op">(</span><span class="va">dat</span><span class="op">)</span><span class="op">==</span><span class="st">"+"</span><span class="op">]</span>
<span class="va">fox2</span><span class="op">$</span><span class="va">dat2</span> <span class="op">&lt;-</span> <span class="va">dat</span><span class="op">[</span><span class="fu">strand</span><span class="op">(</span><span class="va">dat</span><span class="op">)</span><span class="op">==</span><span class="st">"-"</span><span class="op">]</span></code></pre></div>
</div>
<div id="step-2--build-the-gene-model" class="section level3">
<h3 class="hasAnchor">
<a href="#step-2--build-the-gene-model" class="anchor"></a>Step 2. Build the gene model</h3>
<p>The gene model can be built for a given genomic range using <strong>geneModelFromTxdb</strong> function which uses the <strong>TranscriptDb</strong> object as the input.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span>
<span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu">GRanges</span><span class="op">(</span><span class="st">"chr11"</span>, <span class="fu">IRanges</span><span class="op">(</span><span class="fl">122929275</span>, <span class="fl">122930122</span><span class="op">)</span>, strand<span class="op">=</span><span class="st">"-"</span><span class="op">)</span>
<span class="va">trs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/geneModelFromTxdb.html">geneModelFromTxdb</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span>,
                         <span class="va">org.Hs.eg.db</span>,
                         gr<span class="op">=</span><span class="va">gr</span><span class="op">)</span></code></pre></div>
<p>Users can generate a track object with the geneTrack function by inputting a TxDb and a list of gene Entrez IDs. Entrez IDs can be obtained from other types of gene IDs such as gene symbol by using the ID mapping function. For example, to generate a track object given gene FMR1 and human TxDb, refer to the code below.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">entrezIDforFMR1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/get.html">get</a></span><span class="op">(</span><span class="st">"FMR1"</span>, <span class="va">org.Hs.egSYMBOL2EG</span><span class="op">)</span>
<span class="va">theTrack</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/geneTrack.html">geneTrack</a></span><span class="op">(</span><span class="va">entrezIDforFMR1</span>,<span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></code></pre></div>
</div>
<div id="step-3--view-the-tracks" class="section level3">
<h3 class="hasAnchor">
<a href="#step-3--view-the-tracks" class="anchor"></a>Step 3. View the tracks</h3>
<p>Use <strong>viewTracks</strong> function to plot data and annotation information along genomic coordinates. <strong>addGuideLine</strong> or <strong>addArrowMark</strong> can be used to highlight a specific region.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">viewerStyle</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/trackViewerStyle-class.html">trackViewerStyle</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="../reference/trackViewerStyle-class.html">setTrackViewerStyleParam</a></span><span class="op">(</span><span class="va">viewerStyle</span>, <span class="st">"margin"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">.1</span>, <span class="fl">.05</span>, <span class="fl">.02</span>, <span class="fl">.02</span><span class="op">)</span><span class="op">)</span>
<span class="va">trackList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/trackList-class.html">trackList</a></span><span class="op">(</span><span class="va">repA</span>, <span class="va">fox2</span>, <span class="va">trs</span><span class="op">)</span>
<span class="va">vp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/viewTracks.html">viewTracks</a></span><span class="op">(</span><span class="va">trackList</span>, 
                 gr<span class="op">=</span><span class="va">gr</span>, viewerStyle<span class="op">=</span><span class="va">viewerStyle</span>, 
                 autoOptimizeStyle<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="../reference/addGuideLine.html">addGuideLine</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">122929767</span>, <span class="fl">122929969</span><span class="op">)</span>, vp<span class="op">=</span><span class="va">vp</span><span class="op">)</span>
<span class="fu"><a href="../reference/addArrowMark.html">addArrowMark</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>x<span class="op">=</span><span class="fl">122929650</span>, 
                  y<span class="op">=</span><span class="fl">2</span><span class="op">)</span>, <span class="co"># 2 means track 2 from the bottom.</span>
             label<span class="op">=</span><span class="st">"label"</span>,
             col<span class="op">=</span><span class="st">"blue"</span>,
             vp<span class="op">=</span><span class="va">vp</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="trackViewer_files/figure-html/viewTracks-1.png" alt="plot data and annotation information along genomic coordinates" width="768"><p class="caption">
plot data and annotation information along genomic coordinates
</p>
</div>
<p>For more details, please refer <a href="changeTracksStyles.html">vignette changeTracksStyles</a>.</p>
</div>
</div>
<div id="generate-and-edit-an-interactive-plot-using-the-browsetracks-function" class="section level2">
<h2 class="hasAnchor">
<a href="#generate-and-edit-an-interactive-plot-using-the-browsetracks-function" class="anchor"></a>Generate and edit an interactive plot using the browseTracks function</h2>
<p>As shown above, figures produced by trackViewer are highly customizable, allowing users to alter the label, symbol, color, and size with various functions.</p>
<p>For users who prefer to modify the look and feel of a figure interactively, they can use the function <code>browseTracks</code> to draw interactive tracks, leveraging the htmlwidgets package.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/browseTracks.html">browseTracks</a></span><span class="op">(</span><span class="va">trackList</span>, gr<span class="op">=</span><span class="va">gr</span><span class="op">)</span></code></pre></div>
<div class="figure">
<div id="htmlwidget-96e244fdaaeab580ae8b" style="width:576px;height:384px;" class="browseTracks html-widget"></div>
<script type="application/json" data-for="htmlwidget-96e244fdaaeab580ae8b">{"x":{"tracklist":{"uc010rzu.2":{"dat":{"seqnames":"chr11","start":[122928200,122928442,122928960,122929340,122929767,122930181,122930824,122931828,122932033,122932772],"end":[122928441,122928627,122929192,122929538,122929969,122930736,122930951,122932032,122932037,122932901],"width":[242,186,233,199,203,556,128,205,5,130],"strand":["-","-","-","-","-","-","-","-","-","-"],"feature":["utr3","CDS","CDS","CDS","CDS","CDS","CDS","CDS","utr5","utr5"],"id":["unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown"],"exon":["uc010rzu.2_8","uc010rzu.2_8","uc010rzu.2_7","uc010rzu.2_6","uc010rzu.2_5","uc010rzu.2_4","uc010rzu.2_3","uc010rzu.2_2","uc010rzu.2_2","uc010rzu.2_1"],"transcript":["uc010rzu.2","uc010rzu.2","uc010rzu.2","uc010rzu.2","uc010rzu.2","uc010rzu.2","uc010rzu.2","uc010rzu.2","uc010rzu.2","uc010rzu.2"],"gene":["3312","3312","3312","3312","3312","3312","3312","3312","3312","3312"],"symbol":["HSPA8","HSPA8","HSPA8","HSPA8","HSPA8","HSPA8","HSPA8","HSPA8","HSPA8","HSPA8"],"density":[1,1,1,1,1,1,1,1,1,1],"featureLayerID":[1,1,1,1,1,1,1,1,1,1],"fill":["#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6"]},"dat2":[],"style":{"tracktype":"peak","color":"#ADD8E6","NAcolor":"white","breaks":0,"height":0.0666666666666667,"marginTop":0,"marginBottom":0.05,"xscale":{"from":{"x":0.5,"y":0.5,"unit":"npc"},"to":{"x":0.5,"y":0.5,"unit":"npc"},"label":[],"gp":[],"draw":false},"yaxis":{"at":[],"label":false,"gp":[],"draw":true,"main":true},"ylim":[],"ylabpos":"left","ylablas":0,"ylabgp":[]},"ylim":[0,1]},"uc009zbd.2":{"dat":{"seqnames":"chr11","start":[122929941,122930122,122930824,122931301,122931828,122932033,122932772],"end":[122930121,122930736,122930976,122931506,122932032,122932037,122932901],"width":[181,615,153,206,205,5,130],"strand":["-","-","-","-","-","-","-"],"feature":["utr3","CDS","CDS","CDS","CDS","utr5","utr5"],"id":["unknown","unknown","unknown","unknown","unknown","unknown","unknown"],"exon":["uc009zbd.2_5","uc009zbd.2_5","uc009zbd.2_4","uc009zbd.2_3","uc009zbd.2_2","uc009zbd.2_2","uc009zbd.2_1"],"transcript":["uc009zbd.2","uc009zbd.2","uc009zbd.2","uc009zbd.2","uc009zbd.2","uc009zbd.2","uc009zbd.2"],"gene":["3312","3312","3312","3312","3312","3312","3312"],"symbol":["HSPA8","HSPA8","HSPA8","HSPA8","HSPA8","HSPA8","HSPA8"],"density":[1,1,1,1,1,1,1],"featureLayerID":[1,1,1,1,1,1,1],"fill":["#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6"]},"dat2":[],"style":{"tracktype":"peak","color":"#ADD8E6","NAcolor":"white","breaks":0,"height":0.0666666666666667,"marginTop":0,"marginBottom":0.05,"xscale":{"from":{"x":0.5,"y":0.5,"unit":"npc"},"to":{"x":0.5,"y":0.5,"unit":"npc"},"label":[],"gp":[],"draw":false},"yaxis":{"at":[],"label":false,"gp":[],"draw":true,"main":true},"ylim":[],"ylabpos":"left","ylablas":0,"ylabgp":[]},"ylim":[0,1]},"uc009zbc.3":{"dat":{"seqnames":"chr11","start":[122928200,122928442,122928960,122929340,122929767,122930181,122930593,122930824],"end":[122928441,122928627,122929192,122929538,122929969,122930592,122930736,122931094],"width":[242,186,233,199,203,412,144,271],"strand":["-","-","-","-","-","-","-","-"],"feature":["utr3","CDS","CDS","CDS","CDS","CDS","utr5","utr5"],"id":["unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown"],"exon":["uc009zbc.3_6","uc009zbc.3_6","uc009zbc.3_5","uc009zbc.3_4","uc009zbc.3_3","uc009zbc.3_2","uc009zbc.3_2","uc009zbc.3_1"],"transcript":["uc009zbc.3","uc009zbc.3","uc009zbc.3","uc009zbc.3","uc009zbc.3","uc009zbc.3","uc009zbc.3","uc009zbc.3"],"gene":["3312","3312","3312","3312","3312","3312","3312","3312"],"symbol":["HSPA8","HSPA8","HSPA8","HSPA8","HSPA8","HSPA8","HSPA8","HSPA8"],"density":[1,1,1,1,1,1,1,1],"featureLayerID":[1,1,1,1,1,1,1,1],"fill":["#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6"]},"dat2":[],"style":{"tracktype":"peak","color":"#ADD8E6","NAcolor":"white","breaks":0,"height":0.0666666666666667,"marginTop":0,"marginBottom":0.05,"xscale":{"from":{"x":0.5,"y":0.5,"unit":"npc"},"to":{"x":0.5,"y":0.5,"unit":"npc"},"label":[],"gp":[],"draw":false},"yaxis":{"at":[],"label":false,"gp":[],"draw":true,"main":true},"ylim":[],"ylabpos":"left","ylablas":0,"ylabgp":[]},"ylim":[0,1]},"uc001pyp.3":{"dat":{"seqnames":"chr11","start":[122928200,122928442,122929475,122929767,122930181,122930824,122931301,122931828,122932033,122932772],"end":[122928441,122928536,122929538,122929969,122930736,122930976,122931506,122932032,122932037,122932901],"width":[242,95,64,203,556,153,206,205,5,130],"strand":["-","-","-","-","-","-","-","-","-","-"],"feature":["utr3","CDS","CDS","CDS","CDS","CDS","CDS","CDS","utr5","utr5"],"id":["unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown"],"exon":["uc001pyp.3_8","uc001pyp.3_8","uc001pyp.3_7","uc001pyp.3_6","uc001pyp.3_5","uc001pyp.3_4","uc001pyp.3_3","uc001pyp.3_2","uc001pyp.3_2","uc001pyp.3_1"],"transcript":["uc001pyp.3","uc001pyp.3","uc001pyp.3","uc001pyp.3","uc001pyp.3","uc001pyp.3","uc001pyp.3","uc001pyp.3","uc001pyp.3","uc001pyp.3"],"gene":["3312","3312","3312","3312","3312","3312","3312","3312","3312","3312"],"symbol":["HSPA8","HSPA8","HSPA8","HSPA8","HSPA8","HSPA8","HSPA8","HSPA8","HSPA8","HSPA8"],"density":[1,1,1,1,1,1,1,1,1,1],"featureLayerID":[1,1,1,1,1,1,1,1,1,1],"fill":["#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6"]},"dat2":[],"style":{"tracktype":"peak","color":"#ADD8E6","NAcolor":"white","breaks":0,"height":0.0666666666666667,"marginTop":0,"marginBottom":0.05,"xscale":{"from":{"x":0.5,"y":0.5,"unit":"npc"},"to":{"x":0.5,"y":0.5,"unit":"npc"},"label":[],"gp":[],"draw":false},"yaxis":{"at":[],"label":false,"gp":[],"draw":true,"main":true},"ylim":[],"ylabpos":"left","ylablas":0,"ylabgp":[]},"ylim":[0,1]},"uc001pyo.3":{"dat":{"seqnames":"chr11","start":[122928200,122928442,122928960,122929340,122929767,122930181,122930824,122931301,122931828,122932033,122932772],"end":[122928441,122928627,122929192,122929538,122929969,122930736,122930976,122931506,122932032,122932037,122932901],"width":[242,186,233,199,203,556,153,206,205,5,130],"strand":["-","-","-","-","-","-","-","-","-","-","-"],"feature":["utr3","CDS","CDS","CDS","CDS","CDS","CDS","CDS","CDS","utr5","utr5"],"id":["unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown"],"exon":["uc001pyo.3_9","uc001pyo.3_9","uc001pyo.3_8","uc001pyo.3_7","uc001pyo.3_6","uc001pyo.3_5","uc001pyo.3_4","uc001pyo.3_3","uc001pyo.3_2","uc001pyo.3_2","uc001pyo.3_1"],"transcript":["uc001pyo.3","uc001pyo.3","uc001pyo.3","uc001pyo.3","uc001pyo.3","uc001pyo.3","uc001pyo.3","uc001pyo.3","uc001pyo.3","uc001pyo.3","uc001pyo.3"],"gene":["3312","3312","3312","3312","3312","3312","3312","3312","3312","3312","3312"],"symbol":["HSPA8","HSPA8","HSPA8","HSPA8","HSPA8","HSPA8","HSPA8","HSPA8","HSPA8","HSPA8","HSPA8"],"density":[1,1,1,1,1,1,1,1,1,1,1],"featureLayerID":[1,1,1,1,1,1,1,1,1,1,1],"fill":["#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6","#ADD8E6"]},"dat2":[],"style":{"tracktype":"peak","color":"#ADD8E6","NAcolor":"white","breaks":0,"height":0.0666666666666667,"marginTop":0,"marginBottom":0.05,"xscale":{"from":{"x":0.5,"y":0.5,"unit":"npc"},"to":{"x":0.5,"y":0.5,"unit":"npc"},"label":[],"gp":[],"draw":false},"yaxis":{"at":[],"label":false,"gp":[],"draw":true,"main":true},"ylim":[],"ylabpos":"left","ylablas":0,"ylabgp":[]},"ylim":[0,1]},"fox2":{"dat":[],"dat2":[2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,2,2,2,2,2,2,2,2,2,2,2,2,2,1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,3,3,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,5,5,5,4,4,5,5,7,8,8,9,10,11,11,10,10,9,9,10,11,12,13,13,13,13,13,14,15,16,16,16,16,16,15,13,14,14,15,16,16,16,15,14,14,13,11,11,11,11,11,11,11,10,9,8,7,7,7,7,7,6,5,4,4,4,4,4,4,4,3,3,2,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,2,5,6,7,7,8,9,9,11,11,12,12,12,12,12,12,13,14,15,15,16,17,17,17,17,17,18,18,18,18,19,20,19,20,18,16,17,16,16,16,15,14,14,14,16,20,20,21,21,21,20,19,19,20,20,20,20,20,21,22,21,21,21,21,20,18,14,12,13,14,13,13,12,12,12,12,11,11,10,9,9,8,8,8,8,8,7,3,1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,2,3,3,3,3,3,3,3,3,3,3,3,3,3,2,3,3,3,3,3,3,4,4,4,4,4,5,5,5,5,5,5,5,4,4,3,3,3,3,3,3,3,3,3,3,3,3,3,3,2,2,2,2,2,2,1,1,1,1,1,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,2,2,2,2,2,2,3,3,3,3,3,3,3,4,4,5,6,6,7,7,7,7,8,8,8,8,7,7,7,8,8,9,9,9,8,8,9,10,10,10,9,10,10,10,10,10,10,9,9,8,7,7,5,5,5,4,4,4,4,4,4,4,4,4,4,3,3,3,3,3,2,1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,1,2,3,4,5,6,6,6,6,6,6,6,6,6,6,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,6,5,4,3,3,2,2,2,2,2,2,2,2,2,3,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,1,1,1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,2,2,2,2,2,2,2,2,3,3,4,4,5,5,5,5,5,5,5,5,5,6,6,7,8,9,9,9,9,10,10,10,11,13,13,14,14,15,16,17,17,17,16,17,16,16,15,15,16,16,16,16,16,17,18,18,18,17,16,16,16,16,16,15,15,15,13,11,10,10,10,10,9,8,8,8,8,8,9,9,9],"style":{"tracktype":"peak","color":["#000000","#000000"],"NAcolor":"white","breaks":0,"height":0.333333333333333,"marginTop":0,"marginBottom":0.05,"xscale":{"from":{"x":0.5,"y":0.5,"unit":"npc"},"to":{"x":0.5,"y":0.5,"unit":"npc"},"label":[],"gp":[],"draw":false},"yaxis":{"at":[],"label":false,"gp":[],"draw":true,"main":true},"ylim":[],"ylabpos":"left","ylablas":0,"ylabgp":[]},"ylim":[-22,-0]},"repA":{"dat":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0,0,0,0,0.16398,0.16398,0.32796,0.49194,0.65592,0.8199,0.8199,0.8199,1.31184,1.47582,1.6398,1.88577,1.96776,1.96776,3.03362,3.03362,3.11561,3.1976,3.1976,3.1976,3.1976,3.1976,4.0175,4.0175,4.34546,4.34546,4.42745,4.42745,4.42745,4.75541,4.91939,5.08337,5.08337,5.32934,5.41133,5.49332,5.49332,5.41133,5.00138,5.00138,5.00138,4.75541,4.67342,5.08337,5.08337,5.08337,4.34546,4.42745,4.50944,4.67342,5.16536,5.16536,4.26347,4.18148,4.26347,4.75541,4.91939,4.91939,5.00138,5.00138,4.09949,4.09949,4.18148,4.18148,4.09949,4.09949,4.0175,3.60755,3.44357,3.44357,3.44357,3.11561,3.03362,3.03362,3.03362,3.03362,3.1976,3.1976,3.60755,3.36158,3.36158,3.11561,3.11561,3.11561,3.11561,2.70567,1.72179,1.6398,1.6398,1.6398,1.6398,1.14786,1.06587,1.06587,0.8199,0.8199,0.8199,0.8199,0.57393,0.57393,0.57393,0.57393,0.57393,0.57393,0.57393,0.57393,0.57393,0.57393,0.57393,0.57393,0.57393,0.57393,0.57393,0.57393,0,0,0,0.24597,0.32796,0.40995,0.40995,0.40995,0.40995,0.40995,0.8199,0.90189,1.14786,1.14786,1.14786,1.14786,1.6398,2.04975,2.37771,2.86965,3.03362,3.03362,3.03362,3.03362,3.03362,3.03362,3.03362,3.03362,3.03362,3.03362,3.03362,3.03362,3.03362,3.03362,3.1976,3.1976,3.1976,3.1976,3.27959,3.27959,3.27959,3.03362,2.95164,2.86965,2.86965,2.86965,2.86965,2.78766,2.13174,2.21373,2.21373,2.21373,2.21373,2.21373,1.6398,0.65592,0.32796,0.32796,0.32796,0.32796,0.32796,0.32796,0.32796,0.32796,0.32796,0.32796,0.32796,0.32796,0.32796,0.32796,0.32796,0.32796,0.16398,0.16398,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.49194,0.49194,0.49194,0.49194,0.49194,0.49194,0.49194,0.49194,0.49194,0.49194,0.49194,0.49194,0.49194,0.49194,0.49194,0.49194,0.49194,0.49194,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.24597,0.24597,0.32796,0.32796,0.32796,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.40995,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.16398,0.16398,0.24597,0.32796,0.49194,0.57393,0.57393,1.06587,1.06587,1.39383,1.39383,1.47582,1.47582,1.47582,1.47582,1.72179,1.72179,1.72179,1.72179,1.72179,1.72179,1.88577,1.88577,2.21373,2.21373,2.21373,2.21373,2.29572,2.29572,2.29572,2.29572,2.29572,2.29572,2.29572,2.29572,2.29572,2.29572,2.21373,2.21373,2.21373,2.04975,1.96776,1.88577,1.80378,1.80378,1.22985,1.22985,0.90189,0.90189,0.90189,0.90189,0.90189,0.90189,0.65592,0.65592,0.65592,0.65592,0.65592,0.65592,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.08199,0.40995,0.49194,1.06587,1.14786,1.6398,1.80378,1.88577,2.04975,2.04975,2.04975,2.04975,2.04975,2.04975,2.04975,2.04975,2.04975,2.04975,2.04975,2.04975,2.04975,2.04975,2.04975,2.04975,2.04975,2.04975,2.04975,2.04975,2.04975,2.04975,2.04975,2.04975,2.04975,2.04975,2.04975,2.13174,2.13174,2.13174,1.96776,1.80378,1.47582,0.73791,0.73791,0.49194,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0.08199,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"dat2":[],"style":{"tracktype":"peak","color":["#000000","#000000"],"NAcolor":"white","breaks":0,"height":0.333333333333333,"marginTop":0,"marginBottom":0.05,"xscale":{"from":{"x":0.5,"y":0.5,"unit":"npc"},"to":{"x":0.5,"y":0.5,"unit":"npc"},"label":[],"gp":[],"draw":false},"yaxis":{"at":[],"label":false,"gp":[],"draw":true,"main":true},"ylim":[],"ylabpos":"left","ylablas":0,"ylabgp":[]},"ylim":[0,5.49332]}},"type":{"uc010rzu.2":"transcript","uc009zbd.2":"transcript","uc009zbc.3":"transcript","uc001pyp.3":"transcript","uc001pyo.3":"transcript","fox2":"data","repA":"data"},"name":["uc010rzu.2","uc009zbd.2","uc009zbc.3","uc001pyp.3","uc001pyo.3","fox2","repA"],"chromosome":"chr11","start":122929275,"end":122930122,"strand":"*","height":{"uc010rzu.2":0.0666666666666667,"uc009zbd.2":0.0666666666666667,"uc009zbc.3":0.0666666666666667,"uc001pyp.3":0.0666666666666667,"uc001pyo.3":0.0666666666666667,"fox2":0.333333333333333,"repA":0.333333333333333}},"evals":[],"jsHooks":[]}</script><p class="caption">
interactive tracks
</p>
</div>
<p>The videos at <a href="https://youtu.be/lSmeTu4WMlc" class="uri">https://youtu.be/lSmeTu4WMlc</a> and <a href="https://youtu.be/lvF0tnJiHQI" class="uri">https://youtu.be/lvF0tnJiHQI</a> illustrate how to generate and modify an interactive plot. Please note that the interactive feature is only fully implemented in version 1.19.14 or later.</p>
</div>
<div id="plot-multiple-genes-in-one-track" class="section level2">
<h2 class="hasAnchor">
<a href="#plot-multiple-genes-in-one-track" class="anchor"></a>Plot multiple genes in one track</h2>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span>
<span class="va">grW</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parse2GRanges.html">parse2GRanges</a></span><span class="op">(</span><span class="st">"chr11:122,830,799-123,116,707"</span><span class="op">)</span>
<span class="va">ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGeneIDsFromTxDb.html">getGeneIDsFromTxDb</a></span><span class="op">(</span><span class="va">grW</span>, <span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span>
<span class="va">symbols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/get.html">mget</a></span><span class="op">(</span><span class="va">ids</span>, <span class="va">org.Hs.egSYMBOL</span><span class="op">)</span>
<span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/geneTrack.html">geneTrack</a></span><span class="op">(</span><span class="va">ids</span>, <span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span>, 
                   <span class="va">symbols</span>, asList<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="va">optSty</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/optimizeStyle.html">optimizeStyle</a></span><span class="op">(</span><span class="fu"><a href="../reference/trackList-class.html">trackList</a></span><span class="op">(</span><span class="va">repA</span>, <span class="va">fox2</span>, <span class="va">genes</span><span class="op">)</span>, theme<span class="op">=</span><span class="st">"safe"</span><span class="op">)</span>
<span class="va">trackListW</span> <span class="op">&lt;-</span> <span class="va">optSty</span><span class="op">$</span><span class="va">tracks</span>
<span class="va">viewerStyleW</span> <span class="op">&lt;-</span> <span class="va">optSty</span><span class="op">$</span><span class="va">style</span>
<span class="fu"><a href="../reference/viewTracks.html">viewTracks</a></span><span class="op">(</span><span class="va">trackListW</span>, gr<span class="op">=</span><span class="va">grW</span>, viewerStyle<span class="op">=</span><span class="va">viewerStyleW</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="trackViewer_files/figure-html/plotgeneTrack-1.png" alt="Plot multiple genes in one track" width="576"><p class="caption">
Plot multiple genes in one track
</p>
</div>
</div>
</div>
<div id="operators" class="section level1">
<h1 class="hasAnchor">
<a href="#operators" class="anchor"></a>Operators</h1>
<p>If you are interested in drawing a combined track from two input tracks, e.g, adding or substractiong one from the other, then you can try one of the operators such as + and - as showing below.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">newtrack</span> <span class="op">&lt;-</span> <span class="va">repA</span>
<span class="co">## Must keep the same format for dat and dat2</span>
<span class="va">newtrack</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parseWIG.html">parseWIG</a></span><span class="op">(</span><span class="va">newtrack</span>, <span class="st">"chr11"</span>, <span class="fl">122929275</span>, <span class="fl">122930122</span><span class="op">)</span>
<span class="va">newtrack</span><span class="op">$</span><span class="va">dat2</span> <span class="op">&lt;-</span> <span class="va">newtrack</span><span class="op">$</span><span class="va">dat</span>
<span class="va">newtrack</span><span class="op">$</span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="va">fox2</span><span class="op">$</span><span class="va">dat2</span>
<span class="fu"><a href="../reference/trackStyle-class.html">setTrackStyleParam</a></span><span class="op">(</span><span class="va">newtrack</span>, <span class="st">"color"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/viewTracks.html">viewTracks</a></span><span class="op">(</span><span class="fu"><a href="../reference/trackList-class.html">trackList</a></span><span class="op">(</span><span class="va">newtrack</span>, <span class="va">trs</span><span class="op">)</span>, 
           gr<span class="op">=</span><span class="va">gr</span>, viewerStyle<span class="op">=</span><span class="va">viewerStyle</span>, operator<span class="op">=</span><span class="st">"+"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="trackViewer_files/figure-html/viewTracksOperator1-1.png" alt='show data with operator "+"' width="768"><p class="caption">
show data with operator “+”
</p>
</div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/viewTracks.html">viewTracks</a></span><span class="op">(</span><span class="fu"><a href="../reference/trackList-class.html">trackList</a></span><span class="op">(</span><span class="va">newtrack</span>, <span class="va">trs</span><span class="op">)</span>, gr<span class="op">=</span><span class="va">gr</span>, viewerStyle<span class="op">=</span><span class="va">viewerStyle</span>, operator<span class="op">=</span><span class="st">"-"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="trackViewer_files/figure-html/viewTracksOperator2-1.png" alt='show data with operator "-"' width="768"><p class="caption">
show data with operator “-”
</p>
</div>
<p>Alternatively, you can try <strong>GRoperator</strong> before viewing tracks.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">newtrack</span><span class="op">$</span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GRoperator.html">GRoperator</a></span><span class="op">(</span><span class="va">newtrack</span><span class="op">$</span><span class="va">dat</span>, <span class="va">newtrack</span><span class="op">$</span><span class="va">dat2</span>, col<span class="op">=</span><span class="st">"score"</span>, operator<span class="op">=</span><span class="st">"-"</span><span class="op">)</span>
<span class="va">newtrack</span><span class="op">$</span><span class="va">dat2</span> <span class="op">&lt;-</span> <span class="fu">GRanges</span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="../reference/viewTracks.html">viewTracks</a></span><span class="op">(</span><span class="fu"><a href="../reference/trackList-class.html">trackList</a></span><span class="op">(</span><span class="va">newtrack</span>, <span class="va">trs</span><span class="op">)</span>, gr<span class="op">=</span><span class="va">gr</span>, viewerStyle<span class="op">=</span><span class="va">viewerStyle</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="trackViewer_files/figure-html/viewTracksOperator3-1.png" alt='show data with operator "-"' width="768"><p class="caption">
show data with operator “-”
</p>
</div>
</div>
<div id="lolliplot" class="section level1">
<h1 class="hasAnchor">
<a href="#lolliplot" class="anchor"></a>Lolliplot</h1>
<p>The <code>lolliplot</code> function is for the visualization of the methylation/variant/mutation data. For more details, please refer <a href="lollipopPlot.html">vignette lollipopPlot</a>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">trackViewer</span><span class="op">)</span>
<span class="va">features</span> <span class="op">&lt;-</span> <span class="fu">GRanges</span><span class="op">(</span><span class="st">"chr1"</span>, <span class="fu">IRanges</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">501</span>, <span class="fl">1001</span><span class="op">)</span>, 
                                    width<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">120</span>, <span class="fl">400</span>, <span class="fl">405</span><span class="op">)</span>,
                                    names<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"block"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span>,
                    fill <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#FF8833"</span>, <span class="st">"#51C6E6"</span>, <span class="st">"#DFA32D"</span><span class="op">)</span>,
                    height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.02</span>, <span class="fl">0.05</span>, <span class="fl">0.08</span><span class="op">)</span><span class="op">)</span>
<span class="va">SNP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">100</span>, <span class="fl">105</span>, <span class="fl">108</span>, <span class="fl">400</span>, <span class="fl">410</span>, <span class="fl">420</span>, <span class="fl">600</span>, <span class="fl">700</span>, <span class="fl">805</span>, <span class="fl">840</span>, <span class="fl">1400</span>, <span class="fl">1402</span><span class="op">)</span>
<span class="va">sample.gr</span> <span class="op">&lt;-</span> <span class="fu">GRanges</span><span class="op">(</span><span class="st">"chr1"</span>, <span class="fu">IRanges</span><span class="op">(</span><span class="va">SNP</span>, width<span class="op">=</span><span class="fl">1</span>, names<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"snp"</span>, <span class="va">SNP</span><span class="op">)</span><span class="op">)</span>,
                     color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample.int</a></span><span class="op">(</span><span class="fl">6</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">SNP</span><span class="op">)</span>, replace<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>,
                     score <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample.int</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">SNP</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/lolliplot.html">lolliplot</a></span><span class="op">(</span><span class="va">sample.gr</span>, <span class="va">features</span><span class="op">)</span></code></pre></div>
<p><img src="trackViewer_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
</div>
<div id="dandelion-plot" class="section level1">
<h1 class="hasAnchor">
<a href="#dandelion-plot" class="anchor"></a>Dandelion Plot</h1>
<p>Dandelion Plot is used to plot high dense of mutation/methylation data. For more details, please refer <a href="dandelionPlot.html">vignette dandelionPlot</a>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">trackViewer</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span>
<span class="va">methy</span> <span class="op">&lt;-</span> <span class="fu">import</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"methy.bed"</span>, package<span class="op">=</span><span class="st">"trackViewer"</span><span class="op">)</span>, <span class="st">"BED"</span><span class="op">)</span>
<span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu">GRanges</span><span class="op">(</span><span class="st">"chr22"</span>, <span class="fu">IRanges</span><span class="op">(</span><span class="fl">50968014</span>, <span class="fl">50970514</span>, names<span class="op">=</span><span class="st">"TYMP"</span><span class="op">)</span><span class="op">)</span>
<span class="va">trs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/geneModelFromTxdb.html">geneModelFromTxdb</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span>,
                         <span class="va">org.Hs.eg.db</span>,
                         gr<span class="op">=</span><span class="va">gr</span><span class="op">)</span>
<span class="va">features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/RangeTrack-class.html">range</a></span><span class="op">(</span><span class="va">trs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">dat</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/RangeTrack-class.html">range</a></span><span class="op">(</span><span class="va">trs</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">dat</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">features</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">trs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">name</span>, <span class="va">trs</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">name</span><span class="op">)</span>
<span class="va">features</span><span class="op">$</span><span class="va">fill</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"lightblue"</span>, <span class="st">"mistyrose"</span><span class="op">)</span>
<span class="va">features</span><span class="op">$</span><span class="va">height</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">.02</span>, <span class="fl">.04</span><span class="op">)</span>
<span class="fu"><a href="../reference/dandelion.plot.html">dandelion.plot</a></span><span class="op">(</span><span class="va">methy</span>, <span class="va">features</span>, ranges<span class="op">=</span><span class="va">gr</span>, type<span class="op">=</span><span class="st">"pin"</span><span class="op">)</span></code></pre></div>
<p><img src="trackViewer_files/figure-html/unnamed-chunk-3-1.png" width="432"></p>
</div>
<div id="plot-chromatin-interactions-data" class="section level1">
<h1 class="hasAnchor">
<a href="#plot-chromatin-interactions-data" class="anchor"></a>Plot chromatin interactions data</h1>
<p>Plot chromatin interactions heatmap as tracks. For more details, please refer <a href="plotInteractionData.html">vignette plotInteractionData</a>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">InteractionSet</span><span class="op">)</span>
<span class="va">gi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"nij.chr6.51120000.53200000.gi.rds"</span>, package<span class="op">=</span><span class="st">"trackViewer"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span></code></pre></div>
<pre><code>## GInteractions object with 6 interactions and 1 metadata column:
##       seqnames1           ranges1     seqnames2           ranges2 |     score
##           &lt;Rle&gt;         &lt;IRanges&gt;         &lt;Rle&gt;         &lt;IRanges&gt; | &lt;numeric&gt;
##   [1]      chr6 51120000-51160000 ---      chr6 51120000-51160000 |   45.1227
##   [2]      chr6 51120000-51160000 ---      chr6 51160000-51200000 |   35.0006
##   [3]      chr6 51120000-51160000 ---      chr6 51200000-51240000 |   44.7322
##   [4]      chr6 51120000-51160000 ---      chr6 51240000-51280000 |   29.3507
##   [5]      chr6 51120000-51160000 ---      chr6 51280000-51320000 |   38.8417
##   [6]      chr6 51120000-51160000 ---      chr6 51320000-51360000 |   31.7063
##   -------
##   regions: 53 ranges and 0 metadata columns
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">range</span> <span class="op">&lt;-</span> <span class="fu">GRanges</span><span class="op">(</span><span class="st">"chr6"</span>, <span class="fu">IRanges</span><span class="op">(</span><span class="fl">51120000</span>, <span class="fl">53200000</span><span class="op">)</span><span class="op">)</span>
<span class="va">tr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gi2track.html">gi2track</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span>
<span class="va">ctcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"ctcf.sample.rds"</span>,
                            package<span class="op">=</span><span class="st">"trackViewer"</span><span class="op">)</span><span class="op">)</span>
<span class="co">## change the color</span>
<span class="fu"><a href="../reference/trackStyle-class.html">setTrackStyleParam</a></span><span class="op">(</span><span class="va">tr</span>, <span class="st">"breaks"</span>, 
                   <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span>from<span class="op">=</span><span class="fl">0</span>, to<span class="op">=</span><span class="fl">50</span>, by<span class="op">=</span><span class="fl">10</span><span class="op">)</span>, <span class="fl">200</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/trackStyle-class.html">setTrackStyleParam</a></span><span class="op">(</span><span class="va">tr</span>, <span class="st">"color"</span>,
                   <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"lightblue"</span>, <span class="st">"yellow"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/viewTracks.html">viewTracks</a></span><span class="op">(</span><span class="fu"><a href="../reference/trackList-class.html">trackList</a></span><span class="op">(</span><span class="va">ctcf</span>, <span class="va">tr</span>, heightDist<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>, gr<span class="op">=</span><span class="va">range</span>,
           autoOptimizeStyle <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="trackViewer_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
</div>
<div id="ideogram-plot" class="section level1">
<h1 class="hasAnchor">
<a href="#ideogram-plot" class="anchor"></a>Ideogram Plot</h1>
<p>Plot ideograms with a list of chromosomes and a genome.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ideo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadIdeogram.html">loadIdeogram</a></span><span class="op">(</span><span class="st">"hg38"</span>, chrom<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr3"</span>, <span class="st">"chr4"</span>, <span class="st">'chr21'</span>, <span class="st">"chr22"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dataList</span> <span class="op">&lt;-</span> <span class="fu">trim</span><span class="op">(</span><span class="va">ideo</span><span class="op">)</span>
<span class="va">dataList</span><span class="op">$</span><span class="va">score</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">dataList</span><span class="op">$</span><span class="va">gieStain</span><span class="op">)</span><span class="op">)</span>
<span class="va">dataList</span> <span class="op">&lt;-</span> <span class="va">dataList</span><span class="op">[</span><span class="va">dataList</span><span class="op">$</span><span class="va">gieStain</span><span class="op">!=</span><span class="st">"gneg"</span><span class="op">]</span>
<span class="va">dataList</span> <span class="op">&lt;-</span> <span class="fu">GRangesList</span><span class="op">(</span><span class="va">dataList</span><span class="op">)</span>
<span class="fu"><a href="../reference/ideogramPlot.html">ideogramPlot</a></span><span class="op">(</span><span class="va">ideo</span>, <span class="va">dataList</span>, 
            layout<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chr3"</span>, <span class="st">"chr22"</span><span class="op">)</span>, 
                        <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chr4"</span>, <span class="st">"chr21"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="web-application-of-trackviewer" class="section level1">
<h1 class="hasAnchor">
<a href="#web-application-of-trackviewer" class="anchor"></a>Web application of <strong>trackViewer</strong>
</h1>
<p>We created a web application of <strong>trackViewer</strong> (available in 1.19.14 or later) by leveraging the R package <strong>Shiny</strong>. The web application of trackViewer and sample data are available at <a href="https://github.com/jianhong/trackViewer.documentation/tree/master/trackViewerShinyApp" class="uri">https://github.com/jianhong/trackViewer.documentation/tree/master/trackViewerShinyApp</a>. Here is a demo on how to to use the web application at <a href="https://www.nature.com/articles/s41592-019-0430-y#Sec2" class="uri">https://www.nature.com/articles/s41592-019-0430-y#Sec2</a> Supplementary Video 5.</p>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h1>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>R Under development (unstable) (2021-03-18 r80099) Platform: x86_64-pc-linux-gnu (64-bit) Running under: Ubuntu 20.04.2 LTS</p>
<p>Matrix products: default BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so</p>
<p>locale: [1] LC_CTYPE=en_US.UTF-8 LC_NUMERIC=C<br>
[3] LC_TIME=en_US.UTF-8 LC_COLLATE=en_US.UTF-8<br>
[5] LC_MONETARY=en_US.UTF-8 LC_MESSAGES=C<br>
[7] LC_PAPER=en_US.UTF-8 LC_NAME=C<br>
[9] LC_ADDRESS=C LC_TELEPHONE=C<br>
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C</p>
<p>attached base packages: [1] grid parallel stats4 stats graphics grDevices utils<br>
[8] datasets methods base</p>
<p>other attached packages: [1] InteractionSet_1.19.0<br>
[2] httr_1.4.2<br>
[3] VariantAnnotation_1.37.1<br>
[4] Rsamtools_2.7.1<br>
[5] Biostrings_2.59.2<br>
[6] XVector_0.31.1<br>
[7] SummarizedExperiment_1.21.1<br>
[8] MatrixGenerics_1.3.1<br>
[9] matrixStats_0.58.0<br>
[10] org.Hs.eg.db_3.12.0<br>
[11] TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2 [12] GenomicFeatures_1.43.8<br>
[13] AnnotationDbi_1.53.1<br>
[14] Biobase_2.51.0<br>
[15] Gviz_1.35.5<br>
[16] rtracklayer_1.51.5<br>
[17] trackViewer_1.27.11<br>
[18] Rcpp_1.0.6<br>
[19] GenomicRanges_1.43.3<br>
[20] GenomeInfoDb_1.27.8<br>
[21] IRanges_2.25.6<br>
[22] S4Vectors_0.29.11<br>
[23] BiocGenerics_0.37.1</p>
<p>loaded via a namespace (and not attached): [1] colorspace_2.0-0 rjson_0.2.20 ellipsis_0.3.1<br>
[4] rprojroot_2.0.2 biovizBase_1.39.0 htmlTable_2.1.0<br>
[7] base64enc_0.1-3 fs_1.5.0 dichromat_2.0-0<br>
[10] rstudioapi_0.13 bit64_4.0.5 fansi_0.4.2<br>
[13] splines_4.1.0 cachem_1.0.4 knitr_1.31<br>
[16] jsonlite_1.7.2 Formula_1.2-4 cluster_2.1.1<br>
[19] dbplyr_2.1.0 png_0.1-7 grImport_0.9-3<br>
[22] graph_1.69.0 BiocManager_1.30.12 compiler_4.1.0<br>
[25] backports_1.2.1 lazyeval_0.2.2 assertthat_0.2.1<br>
[28] Matrix_1.3-2 fastmap_1.1.0 htmltools_0.5.1.1<br>
[31] prettyunits_1.1.1 tools_4.1.0 gtable_0.3.0<br>
[34] glue_1.4.2 GenomeInfoDbData_1.2.4 dplyr_1.0.5<br>
[37] rappdirs_0.3.3 pkgdown_1.6.1 rhdf5filters_1.3.4<br>
[40] vctrs_0.3.7 debugme_1.1.0 xfun_0.22<br>
[43] stringr_1.4.0 lifecycle_1.0.0 ensembldb_2.15.2<br>
[46] restfulr_0.0.13 XML_3.99-0.6 zlibbioc_1.37.0<br>
[49] scales_1.1.1 BiocStyle_2.19.2 BSgenome_1.59.2<br>
[52] ProtGenerics_1.23.7 ragg_1.1.2 hms_1.0.0<br>
[55] rhdf5_2.35.2 AnnotationFilter_1.15.0 RColorBrewer_1.1-2<br>
[58] yaml_2.2.1 curl_4.3 memoise_2.0.0<br>
[61] gridExtra_2.3 ggplot2_3.3.3 biomaRt_2.47.7<br>
[64] rpart_4.1-15 latticeExtra_0.6-29 stringi_1.5.3<br>
[67] RSQLite_2.2.5 highr_0.8 BiocIO_1.1.2<br>
[70] desc_1.3.0 plotrix_3.8-1 checkmate_2.0.0<br>
[73] filelock_1.0.2 BiocParallel_1.25.5 rlang_0.4.10<br>
[76] pkgconfig_2.0.3 systemfonts_1.0.1 bitops_1.0-6<br>
[79] evaluate_0.14 lattice_0.20-41 Rhdf5lib_1.13.4<br>
[82] purrr_0.3.4 GenomicAlignments_1.27.2 htmlwidgets_1.5.3<br>
[85] bit_4.0.4 tidyselect_1.1.0 magrittr_2.0.1<br>
[88] R6_2.5.0 generics_0.1.0 Hmisc_4.5-0<br>
[91] DelayedArray_0.17.10 DBI_1.1.1 pillar_1.5.1<br>
[94] foreign_0.8-81 survival_3.2-10 KEGGREST_1.31.1<br>
[97] RCurl_1.98-1.3 nnet_7.3-15 tibble_3.1.0<br>
[100] crayon_1.4.1 utf8_1.2.1 BiocFileCache_1.15.1<br>
[103] rmarkdown_2.7 jpeg_0.1-8.1 progress_1.2.2<br>
[106] data.table_1.14.0 Rgraphviz_2.35.0 blob_1.2.1<br>
[109] digest_0.6.27 textshaping_0.3.3 munsell_0.5.0</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Jianhong Ou, Julie Lihua Zhu.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
